rules/terraform_ordered_variable.go

```diff
@@ -0,0 +1,159 @@
+	return true
+}
+
+// Severity returns the rule severity
+func (r *TerraformVariableOrderRule) Severity() tflint.Severity {
+	return tflint.NOTICE
+}
+
+// Link returns the rule reference link
+func (r *TerraformVariableOrderRule) Link() string {
+	return project.ReferenceLink(r.Name())
+}
+
+// Check checks whether the variables are sorted in expected order
+func (r *TerraformVariableOrderRule) Check(runner tflint.Runner) error {
+	path, err := runner.GetModulePath()
+	if err != nil {
+		return err
+	}
+	if !path.IsRoot() {
+		// This rule does not evaluate child modules.
+		return nil
+	}
+
+	body, err := runner.GetModuleContent(&hclext.BodySchema{
+		Blocks: []hclext.BlockSchema{
+			{
+				Type:       "variable",
+				LabelNames: []string{"name"},
+				Body: &hclext.BodySchema{
+					Attributes: []hclext.AttributeSchema{{Name: "default"}},
+				},
+			},
+		},
+	}, &tflint.GetModuleContentOption{IncludeNotCreated: true})
+	if err != nil {
+		return err
+	}
+
+	variables := r.variablesGroupByFile(body)
+	for _, blocks := range variables {
+		if err = r.checkVariableOrder(runner, blocks); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *TerraformVariableOrderRule) variablesGroupByFile(body *hclext.BodyContent) map[string]hclext.Blocks {
+	variables := make(map[string]hclext.Blocks)
+	for _, b := range body.Blocks {
+		variables[b.DefRange.Filename] = append(variables[b.DefRange.Filename], b)
+	}
+	for _, blocks := range variables {
+		sort.Slice(blocks, func(i, j int) bool {
+			return blocks[i].DefRange.Start.Line < blocks[j].DefRange.Start.Line
+		})
+	}
+	return variables
+}
+
+func (r *TerraformVariableOrderRule) checkVariableOrder(runner tflint.Runner, blocks []*hclext.Block) error {
+	requiredVars := r.getSortedVariableNames(blocks, false)
+	optionalVars := r.getSortedVariableNames(blocks, true)
+	sortedVariableNames := append(requiredVars, optionalVars...)
+
+	variableNames := r.getVariableNames(blocks)
+	if reflect.DeepEqual(variableNames, sortedVariableNames) {
+		return nil
+	}
+
+	return runner.EmitIssue(
+		r,
+		"Variables should be sorted in the following order: required(without default value) variables in alphabetical order, optional variables in alphabetical order.",
+		*r.issueRange(blocks),
+	)
+}
+
+func (r *TerraformVariableOrderRule) issueRange(blocks hclext.Blocks) *hcl.Range {
+	requiredVariables := r.getVariables(blocks, false)
+	optionalVariables := r.getVariables(blocks, true)
+
+	for i, b := range requiredVariables {
+		if i > 0 && (b.Labels[0] < requiredVariables[i-1].Labels[0]) {
+			return &b.DefRange
+		}
+	}
+	for i, b := range optionalVariables {
+		if i > 0 && (b.Labels[0] < optionalVariables[i-1].Labels[0]) {
+			return &b.DefRange
+		}
+	}
+
+	firstOptional := optionalVariables[0].DefRange
+	lastRequired := requiredVariables[len(requiredVariables)-1].DefRange
+
+	if firstOptional.Start.Line < lastRequired.Start.Line {
+		return &lastRequired
+	}
+
+	return nil
```

Comment:
nil causes panic. It should return an empty `hcl.Range` if there are no range.

Reply:
Hi, I left this `nil` by intention, please allow me to explain. this `issueRange` is called when we're sure that there is at least one violation inside these blocks, otherwise the caller method `checkVariableOrder` should return a `nil` as error instead. Let's say somehow we've a bug in this `issueRange` method that caused we cannot find any violation, which is wrong, an empty `hcl.Range` will confuse our user, make them wonder what this line 0 issue stands for. In this scenario I think maybe it's better to let it crash so we can receive a bug report that this method somehow has a bug. I'll leave this decision to you, if you prefer an empty `hcl.Range`, I'll be happy to change it in your way.

Reply:
Makes sense. In that case, I think calling `panic` explicitly is correct.
It is better to change the return type to `hcl.Range` instead of `*hcl.Range`, since it is not necessary to return nil.

Reply:
Done. Would you please give another review? Thanks.

Reply:
Sorry for the confusion. I would like to discuss about https://github.com/terraform-linters/tflint-ruleset-terraform/issues/22#issuecomment-1261325415 before restarting the review.